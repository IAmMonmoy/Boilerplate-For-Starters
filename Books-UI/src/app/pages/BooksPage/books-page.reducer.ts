import {
  REQUEST,
  SUCCESS,
  FAILURE,
} from '../../shared/reducers/action-type.util';
import axios from 'axios';

export const ACTIONS_TYPES = {
  FETCH_BOOKS: 'bookPage/FETCH_BOOKS',
};

export interface IBookPageBook {
  bookId: string;
  title: string;
  year: string;
}

const initialState = {
  loading: false,
  errorMessage: null,
  books: [] as ReadonlyArray<IBookPageBook>,
};

const apiUrl = 'books';

export type BookPageState = Readonly<typeof initialState>;

const reducer = (
  state: BookPageState = initialState,
  action,
): BookPageState => {
  switch (action.type) {
    case REQUEST(ACTIONS_TYPES.FETCH_BOOKS):
      return {
        ...state,
        loading: true,
      };
    case SUCCESS(ACTIONS_TYPES.FETCH_BOOKS):
      return {
        ...state,
        loading: false,
        books: action.payload,
      };
    case FAILURE(ACTIONS_TYPES.FETCH_BOOKS):
      return {
        ...state,
        loading: false,
        errorMessage: action.payload,
      };
    default:
      return state;
  }
};

export const getBooks = () => {
  const requestUrl = apiUrl;
  return {
    type: ACTIONS_TYPES.FETCH_BOOKS,
    payload: axios.get(requestUrl),
  };
};

export default reducer;
