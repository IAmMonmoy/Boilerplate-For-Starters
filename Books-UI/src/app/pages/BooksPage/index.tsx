import React, { useEffect } from 'react';
import { Helmet } from 'react-helmet-async';
import Book from '../../components/Book';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import { Container, Row, Col } from 'react-bootstrap';
import { getBooks } from './books-page.reducer';
import { IRootState } from 'app/shared/reducers';
import { connect } from 'react-redux';
export interface IBookPageProps
  extends StateProps,
    DispatchProps,
    RouteComponentProps<any> {}

export const BookPage = ({ getBooks, books }: IBookPageProps) => {
  const history = useHistory();

  useEffect(() => {
    getBooks();
  }, [getBooks]);

  return (
    <Container>
      <Helmet>
        <title>Book Page</title>
        <meta
          name="description"
          content="A React Boilerplate application homepage"
        />
      </Helmet>
      <Row>
        {books &&
          books.map(value => {
            return (
              <Col key={value.bookId}>
                <Book
                  id={value.bookId}
                  title={value.title}
                  year={value.year}
                  onClick={id => {
                    const to = `bookdetail/${id}`;
                    history.push(to);
                  }}
                />
              </Col>
            );
          })}
      </Row>
    </Container>
  );
};

const mapStateToProps = (storeState: IRootState) => ({
  loading: storeState.bookPage.loading,
  books: storeState.bookPage.books,
});

const mapDispatchToProps = { getBooks };

type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;

export default connect(mapStateToProps, mapDispatchToProps)(BookPage);
