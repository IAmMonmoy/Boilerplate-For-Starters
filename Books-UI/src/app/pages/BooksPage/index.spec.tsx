import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import configureStore from 'redux-mock-store';
import promiseMiddleware from 'redux-promise-middleware';
import thunk from 'redux-thunk';
import BooksPage from './index';
import { Provider } from 'react-redux';
const shallowRenderer = createRenderer();

let [getBooks] = new Array(1).fill(jest.fn());

export const testableComponent = component => {
  const mockStore = configureStore([thunk, promiseMiddleware]);
  const store = mockStore({});
  return <Provider store={store}>{component}</Provider>;
};

describe('<BookDeatilPage/>', () => {
  it('should match the snapshot', () => {
    const component = testableComponent(<BooksPage {...shallowSetup()} />);
    const renderedOutput = shallowRenderer.render(component);
    expect(renderedOutput).toMatchSnapshot();
  });
});

function shallowSetup() {
  // Sample props to pass to our shallow render
  const props = {
    books: [
      {
        bookId: '32e959aa-bb2d-11eb-8529-0242ac130003',
        title: 'The Door in the Wall',
        year: 1989,
      },
      {
        bookId: '378bd73a-bb2d-11eb-8529-0242ac130003',
        title: 'Amos Fortune, Free Man',
        year: 1950,
      },
      {
        bookId: '3d634b02-bb2d-11eb-8529-0242ac130003',
        title: 'The Hobbit, Or, There and Back Again',
        year: 2012,
      },
      {
        bookId: '4d634b02-bb2d-11eb-8529-0242ac130003',
        title: 'The Hobbit, Or, There and Back Again',
        year: 1989,
      },
      {
        bookId: '5d634b02-bb2d-11eb-8529-0242ac130003',
        title: 'The Hobbit, Or, There and Back Again',
        year: 1989,
      },
    ],
    match: {} as any,
    done: false,
    getBooks: getBooks,
    loading: false,
    history: {} as any,
    location: {} as any,
  };
  return props;
}
