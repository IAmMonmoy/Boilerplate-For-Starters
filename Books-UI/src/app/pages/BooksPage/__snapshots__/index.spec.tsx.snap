// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDeatilPage/> should match the snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Memo(Connect(BookPage))
    books={
      Array [
        Object {
          "bookId": "32e959aa-bb2d-11eb-8529-0242ac130003",
          "title": "The Door in the Wall",
          "year": 1989,
        },
        Object {
          "bookId": "378bd73a-bb2d-11eb-8529-0242ac130003",
          "title": "Amos Fortune, Free Man",
          "year": 1950,
        },
        Object {
          "bookId": "3d634b02-bb2d-11eb-8529-0242ac130003",
          "title": "The Hobbit, Or, There and Back Again",
          "year": 2012,
        },
        Object {
          "bookId": "4d634b02-bb2d-11eb-8529-0242ac130003",
          "title": "The Hobbit, Or, There and Back Again",
          "year": 1989,
        },
        Object {
          "bookId": "5d634b02-bb2d-11eb-8529-0242ac130003",
          "title": "The Hobbit, Or, There and Back Again",
          "year": 1989,
        },
      ]
    }
    done={false}
    getBooks={[MockFunction]}
    history={Object {}}
    loading={false}
    location={Object {}}
    match={Object {}}
  />
</Context.Provider>
`;
